File: week17/day1_server_neon.js
/*
 * day1_server_neon.js — JS twin of day1_server_neon.py.
 * The full Day 1 server PLUS a query_database tool that runs read-only SQL
 * against a Neon Postgres database — the fix for Failure #3 from the hook.
 *
 * BEFORE RUNNING
 *   npm install                 (adds 'pg' for Postgres)
 *   Set your Neon connection string (NEVER hardcode it):
 *     Mac/Linux:  export NEON_DATABASE_URL="postgresql://user:pass@ep-...neon.tech/neondb?sslmode=require"
 *     Windows:    setx NEON_DATABASE_URL "postgresql://user:pass@ep-...neon.tech/neondb?sslmode=require"
 *
 * RUN:  npx @modelcontextprotocol/inspector@latest -- node day1_server_neon.js
 *
 * EXPOSES tools: read_file, get_weather, wikipedia_summary, web_search, query_database
 *          + resource notes://today + prompt summarize
 */

import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import { z } from "zod";
import pg from "pg";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const server = new McpServer({ name: "Workshop", version: "1.0.0" });

server.registerTool(
  "read_file",
  { description: "Read any text file from the current directory and return its full contents.", inputSchema: { filename: z.string() } },
  async ({ filename }) => {
    const p = path.join(__dirname, filename);
    const text = fs.existsSync(p) ? fs.readFileSync(p, "utf-8") : `File not found: ${filename}`;
    return { content: [{ type: "text", text }] };
  }
);

server.registerTool(
  "get_weather",
  { description: "Get the current weather conditions and temperature for any city in the world.", inputSchema: { city: z.string() } },
  async ({ city }) => {
    try {
      const r = await fetch(`https://wttr.in/${encodeURIComponent(city)}?format=3`);
      return { content: [{ type: "text", text: (await r.text()).trim() }] };
    } catch (e) {
      return { content: [{ type: "text", text: `Weather unavailable: ${e}` }] };
    }
  }
);

server.registerTool(
  "wikipedia_summary",
  {
    description: "Get a short summary of any topic from Wikipedia. Pass a page TITLE (e.g. 'France'), NOT a full question.",
    inputSchema: { topic: z.string() },
  },
  async ({ topic }) => {
    try {
      const r = await fetch("https://en.wikipedia.org/api/rest_v1/page/summary/" + encodeURIComponent(topic),
        { headers: { "User-Agent": "MCP-Workshop/1.0 (educational demo)" } });
      if (!r.ok) throw new Error(`${r.status}`);
      const d = await r.json();
      return { content: [{ type: "text", text: d.extract || "No summary found." }] };
    } catch (e) {
      return { content: [{ type: "text", text: `Wikipedia unavailable: ${e}` }] };
    }
  }
);

server.registerTool(
  "web_search",
  { description: "Search the web for current facts, news, or statistics on any topic.", inputSchema: { query: z.string() } },
  async ({ query }) => {
    try {
      const r = await fetch(`https://api.duckduckgo.com/?q=${encodeURIComponent(query)}&format=json&no_html=1`);
      const d = await r.json();
      const parts = [];
      if (d.AbstractText) parts.push(d.AbstractText);
      for (const t of (d.RelatedTopics || []).slice(0, 3)) if (t.Text) parts.push(t.Text);
      return { content: [{ type: "text", text: parts.join("\n\n") || "No results found." }] };
    } catch (e) {
      return { content: [{ type: "text", text: `Search unavailable: ${e}` }] };
    }
  }
);

// ── NEW TOOL · reach your DATABASE (Neon Postgres) ───────────────────────────
server.registerTool(
  "query_database",
  {
    description:
      "Run a read-only SQL query against our Postgres (Neon) database and return the rows. " +
      "Use this for anything about our OWN data — products, prices, inventory, customers. " +
      "Pass a single SELECT statement, e.g. 'SELECT name, price FROM products'.",
    inputSchema: { sql: z.string() },
  },
  async ({ sql }) => {
    const url = process.env.NEON_DATABASE_URL;
    if (!url) return { content: [{ type: "text", text: "Database not configured: set NEON_DATABASE_URL." }] };
    // Read-only guard: only SELECT, so the model can't change your data.
    if (!sql.trim().toLowerCase().startsWith("select"))
      return { content: [{ type: "text", text: "Only SELECT queries are allowed." }] };

    const client = new pg.Client({ connectionString: url, ssl: { rejectUnauthorized: false } });
    try {
      await client.connect();
      const res = await client.query(sql);
      const rows = res.rows.slice(0, 50);              // cap output so a big table can't flood context
      if (rows.length === 0) return { content: [{ type: "text", text: "No rows." }] };
      const cols = Object.keys(rows[0]);
      const header = cols.join(" | ");
      const body = rows.map((row) => cols.map((c) => String(row[c])).join(" | ")).join("\n");
      return { content: [{ type: "text", text: `${header}\n${body}` }] };
    } catch (e) {
      return { content: [{ type: "text", text: `Database error: ${e}` }] };
    } finally {
      await client.end().catch(() => {});
    }
  }
);

server.registerResource(
  "today-notes",
  "notes://today",
  { title: "Today's notes", description: "Workshop notes the host reads at startup.", mimeType: "text/plain" },
  async (uri) => {
    const p = path.join(__dirname, "notes.txt");
    const text = fs.existsSync(p) ? fs.readFileSync(p, "utf-8") : "No notes found.";
    return { contents: [{ uri: uri.href, text }] };
  }
);

server.registerPrompt(
  "summarize",
  { title: "Summarize a file", description: "Read a file and summarize it.", argsSchema: { filename: z.string() } },
  ({ filename }) => ({
    messages: [{ role: "user", content: { type: "text", text: `Please read the file '${filename}' using the read_file tool, then write a brief, clear summary of its contents.` } }],
  })
);

await server.connect(new StdioServerTransport());

File: week17/day1_server_neon.py
"""
day1_server_neon.py — Day 1 EXERCISE SOLUTION: add a DATABASE tool.

This is a copy of day1_server_full.py with ONE new tool — query_database — that
runs a read-only SQL query against a Neon Postgres database. It's the fix for
Failure #3 from the hook: the model that couldn't see your database now can.

BEFORE RUNNING
    pip install "mcp[cli]" requests psycopg2-binary
    Set your Neon connection string (NEVER hardcode it in the file):
        Mac/Linux:  export NEON_DATABASE_URL="postgresql://user:pass@ep-xxx.neon.tech/neondb?sslmode=require"
        Windows:    setx NEON_DATABASE_URL "postgresql://user:pass@ep-xxx.neon.tech/neondb?sslmode=require"
                    (then reopen the terminal)

RUN:
    python day1_server_neon.py

WHAT IT EXPOSES:
    Tools     : read_file, get_weather, wikipedia_summary, web_search, query_database
    Resource  : notes://today
    Prompt    : summarize
"""

import sys
from pathlib import Path

from mcp.server.fastmcp import FastMCP

mcp = FastMCP("Workshop")


@mcp.tool()
def read_file(filename: str) -> str:
    """Read any text file from the current directory and return its full contents."""
    path = Path(__file__).parent / filename
    if not path.exists():
        return f"File not found: {filename}"
    return path.read_text(encoding="utf-8")


@mcp.tool()
def get_weather(city: str) -> str:
    """Get the current weather conditions and temperature for any city in the world."""
    import requests
    try:
        r = requests.get(f"https://wttr.in/{city}?format=3", timeout=5)
        r.raise_for_status()
        return r.text.strip()
    except Exception as e:
        return f"Weather unavailable: {e}"


@mcp.tool()
def wikipedia_summary(topic: str) -> str:
    """Get a short summary of any topic from Wikipedia.
    Use this for facts, history, definitions, or general knowledge.

    Args:
        topic: A single subject or Wikipedia page title, e.g. 'Cape Verde',
               'France', 'Tel Aviv' — NOT a full question.
    """
    import requests, urllib.parse
    url = "https://en.wikipedia.org/api/rest_v1/page/summary/" + urllib.parse.quote(topic)
    headers = {"User-Agent": "MCP-Workshop/1.0 (educational demo)"}
    try:
        r = requests.get(url, headers=headers, timeout=5)
        r.raise_for_status()
        return r.json().get("extract", "No summary found.")
    except Exception as e:
        return f"Wikipedia unavailable: {e}"


@mcp.tool()
def web_search(query: str) -> str:
    """Search the web for current facts, news, or statistics on any topic.
    Use this when a single Wikipedia page isn't enough — e.g. a country's
    population, a price, or recent events.

    Args:
        query: What to search for, e.g. 'Cape Verde population 2024'.
    """
    from ddgs import DDGS
    try:
        with DDGS() as ddgs:
            results = list(ddgs.text(query, max_results=3))
        if not results:
            return "No results found."
        return "\n\n".join(f"{r['title']}: {r['body']}" for r in results)
    except Exception as e:
        return f"Search unavailable: {e}"


# ── NEW TOOL · reach your DATABASE (Neon Postgres) ───────────────────────────
# Failure #3 from the hook was "the model can't see your database." This is the
# bridge. Neon is just hosted Postgres, so any standard driver works.
#   pip install psycopg2-binary
#   The connection string lives in an ENV VAR, never in the code.
@mcp.tool()
def query_database(sql: str) -> str:
    """Run a read-only SQL query against our Postgres (Neon) database and return
    the rows. Use this for anything about our OWN data — products, prices,
    inventory, customers.

    Args:
        sql: A single SELECT statement, e.g. 'SELECT name, price FROM products'.
    """
    import os
    import psycopg2                       # import inside = isolation (see get_weather)

    url = os.environ.get("NEON_DATABASE_URL")
    if not url:
        return "Database not configured: set NEON_DATABASE_URL to your Neon connection string."

    # Read-only guard: only SELECT is allowed, so the model can't change your data.
    if not sql.strip().lower().startswith("select"):
        return "Only SELECT queries are allowed."

    try:
        conn = psycopg2.connect(url)
        cur = conn.cursor()
        cur.execute(sql)
        cols = [d[0] for d in cur.description]
        rows = cur.fetchmany(50)          # cap the output so a big table can't flood context
        cur.close()
        conn.close()
        if not rows:
            return "No rows."
        header = " | ".join(cols)
        body = "\n".join(" | ".join(str(v) for v in row) for row in rows)
        return f"{header}\n{body}"
    except Exception as e:
        return f"Database error: {e}"


@mcp.resource("notes://today")
def today_notes() -> str:
    """Workshop notes for today. A host app like Goose reads this on startup
    to give the LLM background context without the user having to ask."""
    path = Path(__file__).parent / "notes.txt"
    if not path.exists():
        return "No notes found."
    return path.read_text(encoding="utf-8")


@mcp.prompt(name="summarize")
def summarize_prompt(filename: str) -> str:
    """Prompt template: instructs the LLM to read a file and summarize it."""
    return (
        f"Please read the file '{filename}' using the read_file tool, "
        f"then write a brief, clear summary of its contents."
    )


if __name__ == "__main__":
    if sys.stdin.isatty():
        print("Server starting → http://127.0.0.1:8000/mcp/", file=sys.stderr)
        mcp.run(transport="streamable-http")
    else:
        mcp.run()
